src/System.Buffers.Primitives/System/Buffers/Buffer.cs

```diff
@@ -69,24 +70,19 @@ public Buffer(T[] array, int start, int length)
             if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
```

Comment:
> It looks like: 
`if ((uint)length > array.Length - (uint)start)`  is the same, because if (start > array.Length)  ->  array.Length - (uint)start less than zero and so less than any uint.

I don't understand what you mean. It is the same as what? The two conditions are or'd together.
So, if `((uint)start > (uint)array.Length)` is true, the second condition gets short-circuited. Only if the first condition is false, i.e. start is <= array.Length, would we run the second condition.

Essentially, we need the following behaviour:
```C#
var buffer = new Buffer(new byte[100], 50, 50); // works
var buffer = new Buffer(new byte[100], 101, 1); // it should throw
var buffer = new Buffer(new byte[100], 50, 60); // it should throw, this won't throw if I remove the second condition
```

The checks in the Buffer\<T\> constructor implementation are inspired by Span\<T\>:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L85

> Also, Why casting to uint is necessary?

I believe that is an optimization to try and help the JIT eliminate the redundant bound check. @jkotas, can you confirm?
Although, I don't see any change in the disassembly.
See https://github.com/dotnet/corefxlab/pull/1616#issuecomment-310814785